<a href="https://colab.research.google.com/github/bbchen33/Data-Science/blob/master/Genome_data_wrangling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data wrangling for genome and genes from an unfavored organism - *Clostridium thermocellum*
My fate as a molecular biologist has been mostly dealing with non-model organisms so data wrangling process in this notebook is useful for folks in hell as well.
First, I downloaded the annotated genome from NCBI.

In [1]:
!wget ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/184/925/GCF_000184925.1_ASM18492v1/GCF_000184925.1_ASM18492v1_genomic.gbff.gz

--2020-03-19 02:54:59--  ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/184/925/GCF_000184925.1_ASM18492v1/GCF_000184925.1_ASM18492v1_genomic.gbff.gz
           => ‘GCF_000184925.1_ASM18492v1_genomic.gbff.gz’
Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 130.14.250.10, 2607:f220:41e:250::13
Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|130.14.250.10|:21... connected.
Logging in as anonymous ... Logged in!
==> SYST ... done.    ==> PWD ... done.
==> TYPE I ... done.  ==> CWD (1) /genomes/all/GCF/000/184/925/GCF_000184925.1_ASM18492v1 ... done.
==> SIZE GCF_000184925.1_ASM18492v1_genomic.gbff.gz ... 2437012
==> PASV ... done.    ==> RETR GCF_000184925.1_ASM18492v1_genomic.gbff.gz ... done.
Length: 2437012 (2.3M) (unauthoritative)

GCF_000184925.1_ASM 100%[===================>]   2.32M  1.48MB/s    in 1.6s    

2020-03-19 02:55:05 (1.48 MB/s) - ‘GCF_000184925.1_ASM18492v1_genomic.gbff.gz’ saved [2437012]



In [2]:
!ls #checking that it's in the directory

GCF_000184925.1_ASM18492v1_genomic.gbff.gz  sample_data


In [0]:
!gunzip GCF_000184925.1_ASM18492v1_genomic.gbff.gz # unzip the file

In [4]:
!head -100 GCF_000184925.1_ASM18492v1_genomic.gbff # examine the start of the data

LOCUS       NC_017304            3561619 bp    DNA     circular CON 05-JAN-2020
DEFINITION  Hungateiclostridium thermocellum DSM 1313, complete sequence.
ACCESSION   NC_017304
VERSION     NC_017304.1
DBLINK      BioProject: PRJNA224116
            BioSample: SAMN00713572
            Assembly: GCF_000184925.1
KEYWORDS    GSC:MIGS:2.1; RefSeq.
SOURCE      Hungateiclostridium thermocellum DSM 1313
  ORGANISM  Hungateiclostridium thermocellum DSM 1313
            Bacteria; Firmicutes; Clostridia; Clostridiales;
            Hungateiclostridiaceae; Hungateiclostridium.
REFERENCE   1  (bases 1 to 3561619)
  AUTHORS   Feinberg,L., Foden,J., Barrett,T., Davenport,K.W., Bruce,D.,
            Detter,C., Tapia,R., Han,C., Lapidus,A., Lucas,S., Cheng,J.F.,
            Pitluck,S., Woyke,T., Ivanova,N., Mikhailova,N., Land,M.,
            Hauser,L., Argyros,D.A., Goodwin,L., Hogsett,D. and Caiazza,N.
  TITLE     Complete genome sequence of the cellulolytic thermophile
            Clostridium thermoce

The first many lines are just the information about the genome, when it was sequenced and etc. 

In [5]:
!tail -10 GCF_000184925.1_ASM18492v1_genomic.gbff #examine the end of the data

  3561121 aacacaacaa acacccccaa tccttttttc tgaatcgcta tataaaattt aacctgcgac
  3561181 ttttccgcta aaataaattt tatttaaata aactttgcct tcaatttaac attcaaaaca
  3561241 tgaactataa cgaaaaatac cctattataa attgtaactc actgaaaata ccttatcaat
  3561301 tatatattaa actaattatc aatgtcaaga aactattttt gtggataatt atcaaattaa
  3561361 ttttattttt ttgtggatat cttttcgttt ttttgttgaa tattgcccca aagtctgata
  3561421 atatattaaa tacctgttga taactcgtta ataactcaca atttatcaac agaaattgtt
  3561481 aataactttg ttgacaactt gaacaaaaaa cttttttcgc cttcgtcatt ttactaaaca
  3561541 atacttttta ccactatatt atatgttgtt tttaacaact tatacaaaat tatttcaatt
  3561601 aaatgcttaa gatgccttt
//


The last part of the data contained the actual genome sequence. Things in between are the part that I am interested (gene ID and amino acid sequences)

In [6]:
!sed -n -e '310, 334p' GCF_000184925.1_ASM18492v1_genomic.gbff # showing data in the middle that contain CDS and gene names. 

                     /old_locus_tag="Clo1313_0009"
     CDS             9889..10398
                     /locus_tag="CLO1313_RS00050"
                     /old_locus_tag="Clo1313_0009"
                     /inference="COORDINATES: similar to AA
                     sequence:RefSeq:WP_003513304.1"
                     /note="Derived by automated computational analysis using
                     gene prediction method: Protein Homology."
                     /codon_start=1
                     /transl_table=11
                     /product="DUF4446 family protein"
                     /protein_id="WP_003513304.1"
                     /translation="MSSDVIIDVIELYGPFVLIANAIVTLILIIALFVTNRKLKELKR
                     KYKKFMNGSDNKNIEELLLDNIKLSNDISLKNKEIENRVMSLERDLQRCIQKVGVVRY
                     NAFDDVGSDLSFSIALLDDNDNGVVISGIYSRESSTTYAKKIVNGTSKNPLSAEELQA
                     ISESKKINY"
     gene            10419..11159
                     /locus_tag="CLO1313_RS00055"
                     /old

In this organism, gene names are annotated as "locus_tag" or "old_locus_tag". 

In [0]:
import pandas as pd
table = pd.read_table('GCF_000184925.1_ASM18492v1_genomic.gbff', header = None) # turning the messy texts into a table

In [8]:
table.head()

0
0  LOCUS       NC_017304            3561619 bp   ...
1  DEFINITION  Hungateiclostridium thermocellum D...
2                              ACCESSION   NC_017304
3                            VERSION     NC_017304.1
4                DBLINK      BioProject: PRJNA224116

It looks like from row 0 down to row 130 is just the background information. Let's remove them.

In [9]:
table.iloc[125:138]

0
125                       /organism="Hungateiclostr...
126                            /mol_type="genomic DNA"
127                                 /strain="DSM 1313"
128                       /culture_collection="DSM:...
129                            /db_xref="taxon:637887"
130                            /collection_date="1926"
131                       /collected_by="Viljoen et...
132                          gene            212..1543
133                                       /gene="dnaA"
134                       /locus_tag="CLO1313_RS00010"
135                       /old_locus_tag="Clo1313_0...
136                          CDS             212..1543
137                                       /gene="dnaA"

In [0]:
table = table.iloc[133:]

In [11]:
table.head(30) # making sure the background stuff is gone

0
133                                       /gene="dnaA"
134                       /locus_tag="CLO1313_RS00010"
135                       /old_locus_tag="Clo1313_0...
136                          CDS             212..1543
137                                       /gene="dnaA"
138                       /locus_tag="CLO1313_RS00010"
139                       /old_locus_tag="Clo1313_0...
140                       /inference="COORDINATES: ...
141                       sequence:RefSeq:WP_020457...
142                       /note="Derived by automat...
143                       gene prediction method: P...
144                                     /codon_start=1
145                                   /transl_table=11
146                       /product="chromosomal rep...
147                       /protein_id="WP_003516465.1"
148                       /translation="MNTQLNEIWQK...
149                       AVPAEFNKGILESRYQTLIKNAIKQ...
150                       PLSVLNPKYTFDTFVIGNSNRFAHA...
151                       GHYILEQNSSQRVLYVSSEKFTNEL...
152                       ERTEEEFFHTFNALYEANKQIILSS...
153                       IAILRKKAQLENLTVPNEVIVFIAD...
154                       EALKDILSANKAKVLNCTTIQEAVA...
155                       TEMSLPKIGEEFGGRDHTTVIHACE...
156                         gene            1793..2893
157                       /locus_tag="CLO1313_RS00015"
158                       /old_locus_tag="Clo1313_0...
159                         CDS             1793..2893
160                       /locus_tag="CLO1313_RS00015"
161                       /old_locus_tag="Clo1313_0...
162                               /EC_number="2.7.7.7"

I only want to keep the locus tags and the actual amino acid sequence so I'll delete the rest of the rows.

In [0]:
rows_to_remove_list = ['gene','CDS','/inference','sequence:','/note','gene prediction','/codon_start','transl_table','/product=','protein_id', '/EC_number=']

In [0]:
for item in rows_to_remove_list:
  table = table[~table[0].str.contains(item)]

In [14]:
table.head(50)

0
134                       /locus_tag="CLO1313_RS00010"
135                       /old_locus_tag="Clo1313_0...
138                       /locus_tag="CLO1313_RS00010"
139                       /old_locus_tag="Clo1313_0...
148                       /translation="MNTQLNEIWQK...
149                       AVPAEFNKGILESRYQTLIKNAIKQ...
150                       PLSVLNPKYTFDTFVIGNSNRFAHA...
151                       GHYILEQNSSQRVLYVSSEKFTNEL...
152                       ERTEEEFFHTFNALYEANKQIILSS...
153                       IAILRKKAQLENLTVPNEVIVFIAD...
154                       EALKDILSANKAKVLNCTTIQEAVA...
155                       TEMSLPKIGEEFGGRDHTTVIHACE...
157                       /locus_tag="CLO1313_RS00015"
158                       /old_locus_tag="Clo1313_0...
160                       /locus_tag="CLO1313_RS00015"
161                       /old_locus_tag="Clo1313_0...
171                       /translation="MKIVCSKEQLM...
172                       GNDLELGIECLIDADILEKGSIVLN...
173                       HFELRGMPSDSFPSLPSIEKENMIK...
174                       AGNEITFVSIDGFRMALRKNFNNEG...
175                       NQILFEIGNCKVVSRLLEGEYLNYK...
176                       KYPVKFNIIDDKIIITSNTEIGAVR...
177                       DIYFNSSVGPCTIRPLEGDSFAYMI...
180                       /locus_tag="CLO1313_RS00020"
181                       /old_locus_tag="Clo1313_0...
184                       /locus_tag="CLO1313_RS00020"
185                       /old_locus_tag="Clo1313_0...
194                       /translation="MENIKINTEFI...
195                          ELRRGRKLRTGDIVEINDKKFQIV"
198                       /locus_tag="CLO1313_RS00025"
199                       /old_locus_tag="Clo1313_0...
202                       /locus_tag="CLO1313_RS00025"
203                       /old_locus_tag="Clo1313_0...
212                       /translation="MYIDRILLKNF...
213                       LCASGRSHRTSKDTELVNIDGTGFS...
214                       KKIGNLMGNLLAVIFSPEDILIIKE...
215                       RNMLLKEIQYKRNLLDTLEVWDYKM...
216                       EIMEIKYSPSVDLHDLSNPSEIKNE...
217                       LNLKMFGSQGQQRTSLLSLKLAEIE...
218                       RNVQTFITCTDKELFENRNFGDNLY...
220                       /locus_tag="CLO1313_RS00030"
221                       /old_locus_tag="Clo1313_0...
223                       /locus_tag="CLO1313_RS00030"
224                       /old_locus_tag="Clo1313_0...
233                       /translation="MFLHIGGDRVV...
234                       QTVSSDIPKSFIITETDKKSIIYLS...
235                                                KG"
238                       /locus_tag="CLO1313_RS00035"
239                       /old_locus_tag="Clo1313_0...
242                       /locus_tag="CLO1313_RS00035"

Remove duplicate values. It looks like each locus tag appears twice.

In [0]:
table_unique = table.drop_duplicates()

In [16]:
table_unique.head(20)

0
134                       /locus_tag="CLO1313_RS00010"
135                       /old_locus_tag="Clo1313_0...
148                       /translation="MNTQLNEIWQK...
149                       AVPAEFNKGILESRYQTLIKNAIKQ...
150                       PLSVLNPKYTFDTFVIGNSNRFAHA...
151                       GHYILEQNSSQRVLYVSSEKFTNEL...
152                       ERTEEEFFHTFNALYEANKQIILSS...
153                       IAILRKKAQLENLTVPNEVIVFIAD...
154                       EALKDILSANKAKVLNCTTIQEAVA...
155                       TEMSLPKIGEEFGGRDHTTVIHACE...
157                       /locus_tag="CLO1313_RS00015"
158                       /old_locus_tag="Clo1313_0...
171                       /translation="MKIVCSKEQLM...
172                       GNDLELGIECLIDADILEKGSIVLN...
173                       HFELRGMPSDSFPSLPSIEKENMIK...
174                       AGNEITFVSIDGFRMALRKNFNNEG...
175                       NQILFEIGNCKVVSRLLEGEYLNYK...
176                       KYPVKFNIIDDKIIITSNTEIGAVR...
177                       DIYFNSSVGPCTIRPLEGDSFAYMI...
180                       /locus_tag="CLO1313_RS00020"

In [0]:
table_unique = table_unique.reset_index(drop = True)

In [18]:
table_unique.head(20)

0
0                        /locus_tag="CLO1313_RS00010"
1                        /old_locus_tag="Clo1313_0...
2                        /translation="MNTQLNEIWQK...
3                        AVPAEFNKGILESRYQTLIKNAIKQ...
4                        PLSVLNPKYTFDTFVIGNSNRFAHA...
5                        GHYILEQNSSQRVLYVSSEKFTNEL...
6                        ERTEEEFFHTFNALYEANKQIILSS...
7                        IAILRKKAQLENLTVPNEVIVFIAD...
8                        EALKDILSANKAKVLNCTTIQEAVA...
9                        TEMSLPKIGEEFGGRDHTTVIHACE...
10                       /locus_tag="CLO1313_RS00015"
11                       /old_locus_tag="Clo1313_0...
12                       /translation="MKIVCSKEQLM...
13                       GNDLELGIECLIDADILEKGSIVLN...
14                       HFELRGMPSDSFPSLPSIEKENMIK...
15                       AGNEITFVSIDGFRMALRKNFNNEG...
16                       NQILFEIGNCKVVSRLLEGEYLNYK...
17                       KYPVKFNIIDDKIIITSNTEIGAVR...
18                       DIYFNSSVGPCTIRPLEGDSFAYMI...
19                       /locus_tag="CLO1313_RS00020"

The protein sequence always starts with "/translation" so I can use that to find the index of where protein sequence occurs. 

In [0]:
translation_index = table_unique[table_unique[0].str.contains('/translation')].index

In [20]:
translation_index

Int64Index([    2,    12,    21,    25,    34,    39,    53,    60,    68,
               74,
            ...
            25200, 25217, 25226, 25234, 25249, 25260, 25266, 25275, 25279,
            25284],
           dtype='int64', length=2896)

In [21]:
table_unique[0][12:19].values # looking at the data near the translation index to see the structure of the data

array(['                     /translation="MKIVCSKEQLMEGINVVQKAVPTKATLTILEGILLEAYDNFKMT',
       '                     GNDLELGIECLIDADILEKGSIVLNSKMFGDIVRRLPDSEVLIEVKENNTVIIECDNS',
       '                     HFELRGMPSDSFPSLPSIEKENMIKVSQKAIRDMIRQTLFAVSMEGTRPILTGSLIEC',
       '                     AGNEITFVSIDGFRMALRKNFNNEGFSEFSVVVPAKTLSEIGKILQPVDEDIYIYSSQ',
       '                     NQILFEIGNCKVVSRLLEGEYLNYKSIIPPEYETSVRLRTEDLLSSLERASLITSDEK',
       '                     KYPVKFNIIDDKIIITSNTEIGAVREEIRVEVNGSNMEVGFNPRYFIEALRVIDDELV',
       '                     DIYFNSSVGPCTIRPLEGDSFAYMILPVRINK"'],
      dtype=object)

The locus_tag and old_locus_tag are at the end of the protein sequence so the index of end of protein sequence would be the index of the next protein sequence - 2.

In [0]:
new_df = pd.DataFrame(columns = ['old tag', 'new tag', 'sequence']) # preparing a new dataframe to store "old_locus_tag", "locus_tag" and amino acid sequences.

In [0]:
sequences = []

In [24]:
table_unique[0][2:10].values

array(['                     /translation="MNTQLNEIWQKTLGLLKNELTEISFNTWIKTIDPLSLTGNTINL',
       '                     AVPAEFNKGILESRYQTLIKNAIKQVTFKEYEIAFIVPSQENLNKLTKQTESAGNEDS',
       '                     PLSVLNPKYTFDTFVIGNSNRFAHAAALAVAEAPGKAYNPLFIYGGVGLGKTHLMHAI',
       '                     GHYILEQNSSQRVLYVSSEKFTNELINAIKDNRNEEFRSKYRNIDVLLIDDIQFIAGK',
       '                     ERTEEEFFHTFNALYEANKQIILSSDKPPKEISLEDRLRSRFEWGLIADMQAPDLETR',
       '                     IAILRKKAQLENLTVPNEVIVFIADKIASNIRELEGALNRVIAYSSLTENEITVELAS',
       '                     EALKDILSANKAKVLNCTTIQEAVARYFDIRPEEFKSKKRTRDIAFPRQIAMYLCREL',
       '                     TEMSLPKIGEEFGGRDHTTVIHACEKISEEIESNSETRRAVSEIKRNLLGK"'],
      dtype=object)

Adding the amino acid sequences to the list 'sequences'. 

In [0]:
for i in range(len(translation_index)-1):
  sequences.append(list(table_unique[0][translation_index[i]:translation_index[i+1]-2].values))

In [26]:
len(sequences)

2895

Missing one protein bc the for loop range is -1.

In [27]:
translation_index[-1]

25284

In [29]:
table_unique[0][25284:25590] 

25284                         /translation="MLRTYQPKKRQ...
25285              CONTIG      join(CP002416.1:1..3561619)
25286                                         ORIGIN      
25287            1 attttgtcaa ttattttgta tgcggaaata ttt...
25288           61 taaaatcaac gcaaaatcat aaacaataaa cca...
                               ...                        
25585        17881 attttaatcg gccagtttgt attgttttta aac...
25586        17941 gcggcatctc ttcccagagg gatgtatgat acc...
25587        18001 gtctcaaggt cggcaaaaag caccgcactg tac...
25588        18061 ccgccgactc cgtccactat tttaccgttg aca...
25589        18121 gccaccgttc ttggaccgac tacatcattg aat...
Name: 0, Length: 306, dtype: object

In [0]:
sequences.append(table_unique[0][25284]) # adding the last sequence

In [39]:
len(sequences)

2896

In [40]:
sequences[:3]

[['                     /translation="MNTQLNEIWQKTLGLLKNELTEISFNTWIKTIDPLSLTGNTINL',
  '                     AVPAEFNKGILESRYQTLIKNAIKQVTFKEYEIAFIVPSQENLNKLTKQTESAGNEDS',
  '                     PLSVLNPKYTFDTFVIGNSNRFAHAAALAVAEAPGKAYNPLFIYGGVGLGKTHLMHAI',
  '                     GHYILEQNSSQRVLYVSSEKFTNELINAIKDNRNEEFRSKYRNIDVLLIDDIQFIAGK',
  '                     ERTEEEFFHTFNALYEANKQIILSSDKPPKEISLEDRLRSRFEWGLIADMQAPDLETR',
  '                     IAILRKKAQLENLTVPNEVIVFIADKIASNIRELEGALNRVIAYSSLTENEITVELAS',
  '                     EALKDILSANKAKVLNCTTIQEAVARYFDIRPEEFKSKKRTRDIAFPRQIAMYLCREL',
  '                     TEMSLPKIGEEFGGRDHTTVIHACEKISEEIESNSETRRAVSEIKRNLLGK"'],
 ['                     /translation="MKIVCSKEQLMEGINVVQKAVPTKATLTILEGILLEAYDNFKMT',
  '                     GNDLELGIECLIDADILEKGSIVLNSKMFGDIVRRLPDSEVLIEVKENNTVIIECDNS',
  '                     HFELRGMPSDSFPSLPSIEKENMIKVSQKAIRDMIRQTLFAVSMEGTRPILTGSLIEC',
  '                     AGNEITFVSIDGFRMALRKNFNNEGFSEFSVVVPAKTLSEIGKILQ

The for loop worked. Now I just need to clean up the messy sequences.

In [0]:
import re

def clean_sequence(amino):
  a = ''.join(amino)
  b = re.sub(r"\s+", "", a)
  c = re.sub('/translation=','',b)
  d = re.sub('/locus_tag.*', '', c)
  e = re.sub('"', '', d)
  return e


In [0]:
clean_sequence_list = []
for s in sequences:
  clean_sequence_list.append(clean_sequence(s))

In [45]:
clean_sequence_list[:5]

['MNTQLNEIWQKTLGLLKNELTEISFNTWIKTIDPLSLTGNTINLAVPAEFNKGILESRYQTLIKNAIKQVTFKEYEIAFIVPSQENLNKLTKQTESAGNEDSPLSVLNPKYTFDTFVIGNSNRFAHAAALAVAEAPGKAYNPLFIYGGVGLGKTHLMHAIGHYILEQNSSQRVLYVSSEKFTNELINAIKDNRNEEFRSKYRNIDVLLIDDIQFIAGKERTEEEFFHTFNALYEANKQIILSSDKPPKEISLEDRLRSRFEWGLIADMQAPDLETRIAILRKKAQLENLTVPNEVIVFIADKIASNIRELEGALNRVIAYSSLTENEITVELASEALKDILSANKAKVLNCTTIQEAVARYFDIRPEEFKSKKRTRDIAFPRQIAMYLCRELTEMSLPKIGEEFGGRDHTTVIHACEKISEEIESNSETRRAVSEIKRNLLGK',
 'MKIVCSKEQLMEGINVVQKAVPTKATLTILEGILLEAYDNFKMTGNDLELGIECLIDADILEKGSIVLNSKMFGDIVRRLPDSEVLIEVKENNTVIIECDNSHFELRGMPSDSFPSLPSIEKENMIKVSQKAIRDMIRQTLFAVSMEGTRPILTGSLIECAGNEITFVSIDGFRMALRKNFNNEGFSEFSVVVPAKTLSEIGKILQPVDEDIYIYSSQNQILFEIGNCKVVSRLLEGEYLNYKSIIPPEYETSVRLRTEDLLSSLERASLITSDEKKYPVKFNIIDDKIIITSNTEIGAVREEIRVEVNGSNMEVGFNPRYFIEALRVIDDELVDIYFNSSVGPCTIRPLEGDSFAYMILPVRINK',
 'MENIKINTEFIKLDQFLKWTKTVSMGSEAKLMIRSGLVKVNGEVELRRGRKLRTGDIVEINDKKFQIV',
 'MYIDRILLKNFRNYKDETIKFSKNLNIIYGQNAQGKTNIIEAVFLCASGRSHRTSKDTELVNIDGTGFSVLLDLESSEGRKKIEIDYECGKKKVVKINEIPLKKIG

Woohoo, so much better!

In [0]:
new_df['sequence'] = clean_sequence_list

In [47]:
new_df.head()

old tag new tag                                           sequence
0     NaN     NaN  MNTQLNEIWQKTLGLLKNELTEISFNTWIKTIDPLSLTGNTINLAV...
1     NaN     NaN  MKIVCSKEQLMEGINVVQKAVPTKATLTILEGILLEAYDNFKMTGN...
2     NaN     NaN  MENIKINTEFIKLDQFLKWTKTVSMGSEAKLMIRSGLVKVNGEVEL...
3     NaN     NaN  MYIDRILLKNFRNYKDETIKFSKNLNIIYGQNAQGKTNIIEAVFLC...
4     NaN     NaN  MFLHIGGDRVVPVKNIIAILDMETTTISKDTKDFLAIAEEEGFIQT...

In [48]:
table_unique.head()

0
0                       /locus_tag="CLO1313_RS00010"
1                       /old_locus_tag="Clo1313_0...
2                       /translation="MNTQLNEIWQK...
3                       AVPAEFNKGILESRYQTLIKNAIKQ...
4                       PLSVLNPKYTFDTFVIGNSNRFAHA...

In [52]:
new_df.shape

(2896, 3)

Finding all the indices in the "table_unique" where "locus_tag" occurs.

In [0]:
old_tag_index = table_unique[table_unique[0].str.contains('/old_locus_tag')].index

In [96]:
old_tag_index

Int64Index([    1,    11,    20,    24,    33,    38,    52,    59,    67,
               73,
            ...
            25199, 25216, 25224, 25232, 25247, 25259, 25265, 25273, 25278,
            25283],
           dtype='int64', length=2931)

In [54]:
translation_index

Int64Index([    2,    12,    21,    25,    34,    39,    53,    60,    68,
               74,
            ...
            25200, 25217, 25226, 25234, 25249, 25260, 25266, 25275, 25279,
            25284],
           dtype='int64', length=2896)

Concatenating the "translation_index" and "old_tag_index" and bc I know the "old_tag_index" right before where "translation_index" occurs is the name for that sequence. In other words, index 0 has the locus tag for the amino acid starting at index 2. Index 10 has the locus tag for index 12 and so on. 

In [0]:
import numpy as np
join_array = np.sort(np.concatenate([old_tag_index, translation_index]))

Using np.where to find the value in the concatenated array and -1 for the locus tag index. 

In [0]:
true_old_locus_index = []
for i in translation_index:
  index = join_array[np.where(join_array == i)[0][0] -1]
  true_old_locus_index.append(index)

In [99]:
len(true_old_locus_index) == len(translation_index)

True

They are equal length. It's a good sign.

In [0]:
old_tag_list = []
for locus_tag in table_unique[0][true_old_locus_index]:
  old_tag_list.append(locus_tag)

In [101]:
old_tag_list[:3] # checking how the data look

['                     /old_locus_tag="Clo1313_0001"',
 '                     /old_locus_tag="Clo1313_0002"',
 '                     /old_locus_tag="Clo1313_0003"']

Cleaning the tag name.

In [0]:
def old_tag(tag_name):
  a = re.sub(r"\s+", "", tag_name)
  b = re.sub('/old_locus_tag=','',a)
  c = re.sub('"', '', b)
  return c

In [105]:
old_tag(old_tag_list[0]) #checking if the function works

'Clo1313_0001'

In [106]:
clean_old_tag = []
for tag in old_tag_list:
  clean_old_tag.append(old_tag(tag))
  
print(clean_old_tag[:5])

['Clo1313_0001', 'Clo1313_0002', 'Clo1313_0003', 'Clo1313_0004', 'Clo1313_0005']


In [107]:
len(clean_old_tag)

2896

In [108]:
new_df['old tag'] = clean_old_tag
print(new_df.head()) # looking good

        old tag new tag                                           sequence
0  Clo1313_0001     NaN  MNTQLNEIWQKTLGLLKNELTEISFNTWIKTIDPLSLTGNTINLAV...
1  Clo1313_0002     NaN  MKIVCSKEQLMEGINVVQKAVPTKATLTILEGILLEAYDNFKMTGN...
2  Clo1313_0003     NaN  MENIKINTEFIKLDQFLKWTKTVSMGSEAKLMIRSGLVKVNGEVEL...
3  Clo1313_0004     NaN  MYIDRILLKNFRNYKDETIKFSKNLNIIYGQNAQGKTNIIEAVFLC...
4  Clo1313_0005     NaN  MFLHIGGDRVVPVKNIIAILDMETTTISKDTKDFLAIAEEEGFIQT...


Moving on. Working on the new tag now.

In [0]:
new_tag_index =  table_unique[table_unique[0].str.contains('/locus_tag')].index

In [0]:
join_array2 = np.sort(np.concatenate([new_tag_index, translation_index]))

In [113]:
join_array2

array([    0,     2,    10, ..., 25279, 25282, 25284])

In [0]:
true_new_locus_index = []
for i in translation_index:
  index = join_array2[np.where(join_array2 == i)[0][0] -1]
  true_new_locus_index.append(index)

In [115]:
len(true_new_locus_index) == len(translation_index)

True

In [0]:
new_tag_list = []
for locus_tag in table_unique[0][true_new_locus_index]:
  new_tag_list.append(locus_tag)

In [117]:
new_tag_list[:3]

['                     /locus_tag="CLO1313_RS00010"',
 '                     /locus_tag="CLO1313_RS00015"',
 '                     /locus_tag="CLO1313_RS00020"']

In [0]:
def new_tag(tag_name):
  a = re.sub(r"\s+", "", tag_name)
  b = re.sub('/locus_tag=','',a)
  c = re.sub('"', '', b)
  return c

In [119]:
new_tag(new_tag_list[0])

'CLO1313_RS00010'

In [120]:
clean_new_tag = []
for tag in new_tag_list:
  clean_new_tag.append(new_tag(tag))
  
print(clean_new_tag[:5])

['CLO1313_RS00010', 'CLO1313_RS00015', 'CLO1313_RS00020', 'CLO1313_RS00025', 'CLO1313_RS00030']


In [121]:
new_df['new tag'] = clean_new_tag
print(new_df.head())

        old tag  ...                                           sequence
0  Clo1313_0001  ...  MNTQLNEIWQKTLGLLKNELTEISFNTWIKTIDPLSLTGNTINLAV...
1  Clo1313_0002  ...  MKIVCSKEQLMEGINVVQKAVPTKATLTILEGILLEAYDNFKMTGN...
2  Clo1313_0003  ...  MENIKINTEFIKLDQFLKWTKTVSMGSEAKLMIRSGLVKVNGEVEL...
3  Clo1313_0004  ...  MYIDRILLKNFRNYKDETIKFSKNLNIIYGQNAQGKTNIIEAVFLC...
4  Clo1313_0005  ...  MFLHIGGDRVVPVKNIIAILDMETTTISKDTKDFLAIAEEEGFIQT...

[5 rows x 3 columns]


In [0]:
pd.set_option('display.max_columns', None) # to expand the columns

In [123]:
new_df.head()

old tag          new tag  \
0  Clo1313_0001  CLO1313_RS00010   
1  Clo1313_0002  CLO1313_RS00015   
2  Clo1313_0003  CLO1313_RS00020   
3  Clo1313_0004  CLO1313_RS00025   
4  Clo1313_0005  CLO1313_RS00030   

                                            sequence  
0  MNTQLNEIWQKTLGLLKNELTEISFNTWIKTIDPLSLTGNTINLAV...  
1  MKIVCSKEQLMEGINVVQKAVPTKATLTILEGILLEAYDNFKMTGN...  
2  MENIKINTEFIKLDQFLKWTKTVSMGSEAKLMIRSGLVKVNGEVEL...  
3  MYIDRILLKNFRNYKDETIKFSKNLNIIYGQNAQGKTNIIEAVFLC...  
4  MFLHIGGDRVVPVKNIIAILDMETTTISKDTKDFLAIAEEEGFIQT...

With the nice and clean table available, one can ask a question such as what the new tag and the amino acid sequence of Clo1313_0396 are?

In [125]:
new_df[new_df['old tag'] == 'Clo1313_0396']

old tag          new tag  \
384  Clo1313_0396  CLO1313_RS02005   

                                              sequence  
384  MTSEEIAKLCGVSRATVSRVINNSPNVKEETRQKILAVIKEKNYVP...

From previous experience, I found that the sequence could contain numbers and alphabets that weren't supposed to be amino acids so I will remove them with the function below.

In [0]:
def clean_sequence(seq):
  new_seq = ''
  amino_acids = ['A', 'C', 'D', 'E', 'F', 'G', 
                'H', 'I', 'K', 'L', 'M', 'N', 
                'P', 'Q', 'R', 'S', 'T', 'V',
                'V', 'W', 'Y']
  for s in seq:
    if s in amino_acids:
      new_seq += s
  return new_seq

In [0]:
new_df['sequence'] = new_df['sequence'].apply(lambda x:clean_sequence(x))

In [151]:
round(new_df['sequence'].str.len().sum()/2896, 2)

333.64

The average protein sequence is 333.64 amino acids in this organism. 

One can also ask: "What is the most abundant amino acid in this organism?"

In [0]:
complete_aa = ''
for i in range(len(new_df)):
  complete_aa += new_df.sequence[i]

In [153]:
len(complete_aa)/2896

333.6401933701657

In [158]:
from collections import Counter
x = Counter(complete_aa)
sorted(x.items(), key = lambda i: i[1], reverse = True)

[('L', 83715),
 ('I', 82157),
 ('K', 77141),
 ('E', 73002),
 ('V', 70068),
 ('G', 65504),
 ('A', 62302),
 ('S', 58702),
 ('D', 55258),
 ('N', 51802),
 ('T', 48782),
 ('F', 41405),
 ('R', 40978),
 ('Y', 40321),
 ('P', 33248),
 ('M', 24368),
 ('Q', 24324),
 ('H', 13524),
 ('C', 11324),
 ('W', 8297)]

Leucine is the most common amino acid in this organism. 